### PR TITLE
Issue due date

### LINKED ISSUE #2533: Due Date for issues
It would be nice if one would be able to set a due date per issue, similar to how milestones can have a due date.

Sometimes you only need one feature to be done until a given date, so there is no need to create an extra milestone only for this.

### PR DESCRIPTION
**Pull request by kolaente**
Fixes #2533.
~~Blocked by go-gitea/go-sdk#103.~~

This PR adds the ability to add due dates to issues. This works similar to milestones, but for single issues. Due date can be managed via the api.
Due dates are shown in the issue list, overdue issues are marked in red.

## Screenshots
bildschirmfoto von 2018-04-14 17-19-43
bildschirmfoto von 2018-04-14 17-19-55
bildschirmfoto von 2018-04-14 17-21-12

## TODO
* [x] Update `go-sdk` once its merged

### CODE REVIEW

File: models/issue.go
Comment (lunny) [inline models/issue.go]: It seems this field is unnecessary.

File: models/issue.go
Comment (lunny) [inline models/issue.go]: this should be `err = updateIssue(sess, issue)` and need `sess.Commit()`

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/issue.go
Comment (jonasfranz) [reply models/issue.go]: You could use something like `.DeadlineUnix.FormatShort`.

Comment (kolaente) [reply models/issue.go]: Thanks, I didn't know that function exists.

On another note, what do you think about putting date formats in the corresponding locale files meaning we could have different date formats for every locale?

File: models/issue_comment.go
Comment (jonasfranz) [inline models/issue_comment.go]: I think that you could pack all of these `create...DeadlineComment` functions into one function with an additional parameter `type` (please also check if the type is a valid deadline type).

File: modules/auth/repo_form.go
Comment (jonasfranz) [inline modules/auth/repo_form.go]: Replace `MaxSize` with `Size` since a `yyyy-mm-dd` date has always 10 characters.

File: templates/repo/issue/view_content/comments.tmpl
Comment (jonasfranz) [inline templates/repo/issue/view_content/comments.tmpl:199]: Please remove `|Safe` since it is not required (no *safe* HTML content inside)

File: templates/repo/issue/view_content/comments.tmpl
Comment (jonasfranz) [inline templates/repo/issue/view_content/comments.tmpl]: Please remove `|Safe` since it is not required (no *safe* HTML content inside)

File: templates/repo/issue/view_content/comments.tmpl
Comment (jonasfranz) [inline templates/repo/issue/view_content/comments.tmpl:219]: Please remove `|Safe` since it is not required (no *safe* HTML content inside)

File: templates/repo/issue/view_content/sidebar.tmpl
Comment (jonasfranz) [inline templates/repo/issue/view_content/sidebar.tmpl]: Replace `placeholder='...'` with `placeholder="..."`

File: options/locale/locale_en-US.ini
Comment (jonasfranz) [inline options/locale/locale_en-US.ini:745]: I think that
> format yyyy-mm-dd

could be confusing for users from non-english countries (like Germany) because modern browsers use for the UI the local date format (in Germany: dd.mm.yyyy) and as value `yyyy-mm-dd`. (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date#Value)

**Review by jonasfranz — CHANGES_REQUESTED** [review]

File: options/locale/locale_en-US.ini
Comment (kolaente) [reply options/locale/locale_en-US.ini:745]: Yes, but that value is used as a placeholder for the input. If the browser supports `input type date`, the browser does what you described thus not showing the placeholder at all (because there's already something inside that field).

Some screenshots with an empty input:

Firefox:
bildschirmfoto von 2018-04-16 18-42-51

Chromium:
bildschirmfoto von 2018-04-16 18-43-22

The format doesnt really seem to be an issue here.

Also, currently the date is not localized anywhere, so if we'd start doing that here we should do it everywhere which is totally out of scope for this pr.

File: templates/repo/issue/view_content/comments.tmpl
Comment (kolaente) [reply templates/repo/issue/view_content/comments.tmpl:199]: If I do that, it looks like that:
bildschirmfoto von 2018-04-16 18-47-55

So I'm leaving it. (same for the other ones)

File: models/issue_comment.go
Comment (kolaente) [reply models/issue_comment.go]: Done, good idea!

File: modules/auth/repo_form.go
Comment (kolaente) [reply modules/auth/repo_form.go]: Thanks, I didn't know `Size` exists.

File: options/locale/locale_en-US.ini
Comment (jonasfranz) [reply options/locale/locale_en-US.ini:745]: If the placeholder isn't shown at all it could be removed to reduce translation effort.

File: templates/repo/issue/view_content/comments.tmpl
Comment (jonasfranz) [reply templates/repo/issue/view_content/comments.tmpl:199]: Fair enough

File: models/issue.go
Comment (lunny) [inline models/issue.go]: why not just change IsOverDue as a method of Issue?

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: Good idea, I've changed it.

File: models/issue.go
Comment (lunny) [inline models/issue.go]: how about `return util.TimeStampNow() >= issue.DeadlineUnix`

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/issue.go
Comment (lunny) [inline routers/api/v1/repo/issue.go]: This line could be removed I think.

File: routers/api/v1/repo/pull.go
Comment (lunny) [inline routers/api/v1/repo/pull.go]: The same as above.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: Fixed.

File: routers/api/v1/repo/issue.go
Comment (kolaente) [reply routers/api/v1/repo/issue.go]: Seems logic, I removed that.

**Comment by jonasfranz** [comment]
LGTM

File: routers/api/v1/repo/issue.go
Comment (Morlinest) [inline routers/api/v1/repo/issue.go]: Missing `return`?

File: routers/api/v1/repo/pull.go
Comment (Morlinest) [inline routers/api/v1/repo/pull.go]: Missing `return`?

File: models/issue.go
Comment (Morlinest) [inline models/issue.go]: Please correct comment...

File: models/issue.go
Comment (Morlinest) [inline models/issue.go]: 1. Do you really need to load whole issue?
1. You should load it only if needed (above L1535)

File: models/issue.go
Comment (Morlinest) [inline models/issue.go]: You can change this to `func UpdateIssueDeadline(issue *Issue, DeadlineUnix util.TimeStamp, doer *User) error {`. Then you can remove `actualIssue`.

**Review by Morlinest — CHANGES_REQUESTED** [review]

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: How? Do you want me to write something like `UpdateIssueDeadline updates an issue deadline`?

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: 1. No, technically I only need `DeadlineUnix`, but I don't think this has a huge impact if I either load the full issue or only `DeadlineUnix`. Is that even possible with xorm?
2. I'd need it in both cases (L1535 & L1529)

File: routers/api/v1/repo/issue.go
Comment (kolaente) [reply routers/api/v1/repo/issue.go]: Probably. I've added one (same for pull.go).

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: But wouldn't that mean to get the actual deadline before calling the function?

**Review by lunny — APPROVED** [review]

File: models/issue.go
Comment (Morlinest) [reply models/issue.go]: That's better. And does it really update only deadline? I think it updates all given fields right now.

File: models/issue.go
Comment (Morlinest) [reply models/issue.go]: 1. It should be possible.
1. You are right, i missed that one line.

File: models/issue.go
Comment (Morlinest) [reply models/issue.go]: Maybe something like this?
```
	var deadline util.TimeStamp
	if form.Deadline != nil && !form.Deadline.IsZero() {
		deadline  = util.TimeStamp(form.Deadline.Unix())
	}
	if err := models.UpdateIssueDeadline(issue, deadline, ctx.User); err != nil {
		ctx.Error(500, "UpdateIssueDeadline", err)
	}
```

File: models/issue.go
Comment (Morlinest) [reply models/issue.go]: It also creates comments. I am not sure if it is good. Maybe better to add comments directly in udate/remove deadline function?

Comment (kolaente) [reply models/issue.go]: > That's better. And does it really update only deadline? I think it updates all given fields right now.

Well, it updates all fields to its old value + issueDeadline to the new value. Although I think xorm does some magic here and only updates the fields which did not change.

> It also creates comments. I am not sure if it is good. Maybe better to add comments directly in udate/remove deadline function?

This is the function which updates/removes the deadline...

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: I've changed it to only get and update `deadline_unix`.

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: But I'd still need to check if the issue deadline hasn't changed meanwhile.

File: models/issue.go
Comment (Morlinest) [reply models/issue.go]: So the only diff between this function and `updateIssue` is that this func creates comments. Function doesn't know what are the fields of `issue`, who and why set it to given values, only caller knows it (function doesn't know context about when it is called, I can call it from anywhere). If I look at implementation, more meaningful explanation is "UpdateIssueDeadline creates deadline comment and updates issue".

I don't think Xorm is so magical. As I know, it can ignore default values, but doesn't know what is updated. But you can check it by turning on sql logging so we can be sure about that.

Sry, I should be more concrete about "update/remove" functions. I meant `UpdateDeadline` and `RemoveDeadline`. But as there is not `xorm.Session` it can not be called from there directly.


Edit: OK, you updated it in meanwhile so this comment is not 100% relevant now :)

**Comment by kolaente** [comment]
@Morlinest  could you please re-review/approve the changes I made?

File: models/issue.go
Comment (Morlinest) [inline models/issue.go]: We can move this if/else to `createDeadlineComments` to be consistent (no other create*Comment function takes CommentType as argument)
`createDeadlineComment(sess, doer, repo, issue, issue.DeadlineUnix, deadlineUnix)`

File: models/issue_comment.go
Comment (Morlinest) [inline models/issue_comment.go]:
```
func createDeadlineComment(e *xorm.Session, doer *User, repo *Repository, issue *Issue, oldDeadline, newDeadline util.TimeStamp) (*Comment, error) {
	commentType := CommentTypeAddedDeadline
	deadline := newDeadline 

	if newDeadline == 0 {
		deadline = oldDeadline
		commentType := CommentTypeRemovedDeadline
	} else if oldDeadline != 0 {
		commentType := CommentTypeModifiedDeadline
	}

	content := deadline.Format("2006-01-02")

	return createComment(e, &CreateCommentOptions{
		Type:    commentType,
		Doer:    doer,
		Repo:    repo,
		Issue:   issue,
		Content: content,
	})
}
```

It would be great if content is more specific. When deadline is modified it should show both dates.

**Review by Morlinest — CHANGES_REQUESTED** [review]
Few (all) related changes.

File: routers/api/v1/repo/issue.go
Comment (kolaente) [reply routers/api/v1/repo/issue.go]: What do I need the repo for?

**Comment by kolaente** [comment]
I found something interesting in `issue_comment.go:426`, it updates the issue when I create a comment. This destroys my routine, as it updates the issue back to the old deadline right after I updated it to the new one.
This seems like a bug, should I fix that in a seperate pr?

File: routers/api/v1/repo/issue.go
Comment (Morlinest) [reply routers/api/v1/repo/issue.go]: for `createDeadlineComment`

Comment (kolaente) [reply routers/api/v1/repo/issue.go]: You're right, but isn't the repo already available in `issue.Repo`?

**Comment by Morlinest** [comment]
For me - yes, it looks like bug. Now you see reason why I wanted you to change comment/code, to prevent this kind of bugs.

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: Yes. Fixed that.

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: You're right, I haven't thought of that.

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: Done.

File: models/issue_comment.go
Comment (kolaente) [reply models/issue_comment.go]: Sure, but I rather won't add another row to the issue comment table as there are already too much. Someone said we should redo that some day, so I'll probably wait for that to do that...

Or do you have an idea on how to add both dates in one field?

Comment (Morlinest) [reply models/issue_comment.go]: I meant if we can use `Content` and add both dates there.

Comment (kolaente) [reply models/issue_comment.go]: Yes, sure. I'll figure something out.

Comment (kolaente) [reply models/issue_comment.go]: Done.

**Comment by kolaente** [comment]
@Morlinest I see, I've created a pr to fix that: #3855

File: models/issue.go
Comment (Morlinest) [inline models/issue.go:1559]: If you want transaction, you have to call `Begin()` first (from xorm doc):
```
if err := session.Begin(); err != nil {
    // if returned then will rollback automatically
    return err
}
```

If not, this is not needed and you can just `return nil`.

File: models/issue.go
Comment (Morlinest) [inline models/issue.go]: We can also use `UpdateIssueCols` or `updateIssueCols` here (I did not know about that func before). It also calls issue indexer so it is more "safe" for the future changes.

`UpdateIssueCols(&Issue{ID: issue.ID, DeadlineUnix: deadlineUnix}, "deadline_unix")`

or

`updateIssueCols(sess, &Issue{ID: issue.ID, DeadlineUnix: deadlineUnix}, "deadline_unix")`

What do you think?

**Review by Morlinest — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
need @JonasFranzDEV @Morlinest 's approvals.

File: models/issue.go
Comment (kolaente) [reply models/issue.go:1543]: I need that session to create the comment.

File: models/issue.go
Comment (kolaente) [reply models/issue.go:1559]: Sure, I'll add that.

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: Sure, I didn't knew that function either.

File: models/issue.go
Comment (Morlinest) [inline models/issue.go]: `defer sess.Close()` should be after `sess := x.NewSession()` and before `sess.Begin()`

File: routers/api/v1/repo/issue.go
Comment (Morlinest) [inline routers/api/v1/repo/issue.go:275]: nit: Use `var deadlineUnix util.TimeStamp` or `var deadline util.TimeStamp`, not both as it has same meaning (same in _repo/issue.go_ and _repo/pull.go_).

File: models/issue.go
Comment (kolaente) [reply models/issue.go]: Sure, fixed.

File: routers/api/v1/repo/issue.go
Comment (kolaente) [reply routers/api/v1/repo/issue.go:275]: Sure, I fixed that as well.

File: routers/repo/issue.go
Comment (Morlinest) [inline routers/repo/issue.go]: I found there is authorization middleware, can you use that instead? Like in `m.Post("/status", reqRepoWriter, repo.UpdateIssueStatus)`. It calls `RequireRepoWriter()` func.

Comment (kolaente) [reply routers/repo/issue.go]: Done.

**Comment by jonasfranz** [comment]
@lunny I already approved it here: https://github.com/go-gitea/gitea/pull/3794#issuecomment-383282018

**Review by jonasfranz — APPROVED** [review]

File: routers/routes/routes.go
Comment (Morlinest) [inline routers/routes/routes.go]: For consistency, this endpoint should be called `/delete`.

File: templates/repo/issue/view_content/sidebar.tmpl
Comment (Morlinest) [inline templates/repo/issue/view_content/sidebar.tmpl]: This can/should be standard link (`<a href="{{$.RepoLink}}/issues/{{.Issue.Index}}/deadline/remove">`), right?

**Review by Morlinest — CHANGES_REQUESTED** [review]
Should be last changes. There are another two things I don' like. But one is not related directly only to this issue (usage of `ctx.Writen()` and `ctx.HasError()`) and second is about better UI (but that can be updated in future, right now I have not better idea).

File: templates/repo/issue/view_content/sidebar.tmpl
Comment (kolaente) [reply templates/repo/issue/view_content/sidebar.tmpl]: No, the js function sends a post request.

File: routers/routes/routes.go
Comment (kolaente) [reply routers/routes/routes.go]: Fixed that.

File: templates/repo/issue/view_content/sidebar.tmpl
Comment (kolaente) [reply templates/repo/issue/view_content/sidebar.tmpl]: I'll send a seperate pr after this one is merged which will add a seperate api endpoint for deadlines which then will be used for the update the deadline from the UI.

Comment (Morlinest) [reply templates/repo/issue/view_content/sidebar.tmpl]: Good to know you will improve this solution and add api endpoint. Until then, you should not simulate api endpoint by using js and standard page, ignoring anything it can return. For now, there is not reason to use js, you can add it later, in another PR when api would be ready for it.

Comment (kolaente) [reply templates/repo/issue/view_content/sidebar.tmpl]: I've modifed the endpoint, deleting a deadline is now done via GET.

**Review by Morlinest — APPROVED** [review]
gj

**Comment by kolaente** [comment]
@Morlinest thanks!

**Review by lafriks — CHANGES_REQUESTED** [review]
@Morlinest @kolaente Wanted to already merge but there one thing needs to be changed back as it was in previous commits. For security reasons all actions that update database must be done using post method, so that CSRF can be verified and user could not be fooled into actions that he did not want (either by giving him url or redirecting him to such url)

**Comment by kolaente** [comment]
@lafriks OK, I'll change it back to how it was before.

**Comment by kolaente** [comment]
@lafriks Done.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
@Morlinest about doing `post` for removing due date?

**Pull request merged**
